src/Api.js

```diff
@@ -89,13 +95,97 @@ class Api {
 	}
 
 	/**
-	 * Get the WhoColor data for a given wiki page.
+	 * Extract token and user IDs from a WikiWho span element with `id='token-X'` and
+	 * `class='token-editor-Y'` attributes.
 	 *
-	 * @param  {string} wikiUrl URL of the wiki page that we want to analyze.
+	 * @param {HTMLElement} element
+	 * @return {int[]|boolean[]} A two-element array: 0 = token_id, 1 = user_id.
+	 */
+	getIdsFromElement( element ) {
+		const out = [ false, false ],
+			tokenMatches = element.getAttribute( 'id' ).match( /token-(\d+)/ ),
+			editorMatches = element.getAttribute( 'class' ).match( /token-editor-(\d+)/ );
+		if ( tokenMatches && tokenMatches[ 1 ] ) {
+			out[ 0 ] = parseInt( tokenMatches[ 1 ] );
+		}
+		if ( editorMatches && editorMatches[ 1 ] ) {
+			out[ 1 ] = parseInt( editorMatches[ 1 ] );
+		}
+		return out;
+	}
+
+	/**
+	 * Get the WikiWho replacement for `.mw-parser-output` HTML.
+	 * @return {string}
+	 */
+	getReplacementHtml() {
+		return this.results.extended_html;
+	}
+
+	/**
+	 * Get user and revision information for a given token.
+	 *
+	 * @param {int} tokenId
+	 * @return {{revisionId: *, score: *, userId: *, username: *, revisionTime: *}|boolean}
+	 */
+	getTokenInfo( tokenId ) {
+		let revId, revision, username, score;
+
+		// Get the token information. results.tokens structure:
+		// [ [ conflict_score, str, o_rev_id, in, out, editor/class_name, age ], ... ]
+		// e.g. Array(7) [ 0, "indicate", 769691068, [], [], "18201938", 76652371.587203 ]
+		const token = this.results.tokens[ tokenId ] || false;
+		if ( !token ) {
+			return false;
+		}
+
+		// Get revision information. results.revisions structure:
+		// { rev_id: [ timestamp, parent_rev_id, user_id, editor_name ], ... }
+		// e.g. Array(4) [ "2017-03-11T02:12:47Z", 769315355, "18201938", "Biogeographist" ]
+		revId = token[ 2 ];
+		revision = this.results.revisions[ revId ];
+		username = revision[ 3 ];
+
+		// Get the user's edit score (percentage of content edited).
+		// results.present_editors structure:
+		// [ [ username, user_id, score ], ... ]
+		for ( let i = 0; i < this.results.present_editors.length; i++ ) {
+			if ( this.results.present_editors[ i ][ 0 ] === username ) {
+				score = parseFloat( this.results.present_editors[ i ][ 2 ] ).toFixed( 1 );
+				break;
+			}
+		}
+
+		// Put it all together.
+		return {
+			username: username,
+			userId: token[ 5 ],
+			revisionId: revId,
+			revisionTime: new Date( revision[ 0 ] ),
+			score: score
+		};
+	}
+
+	/**
+	 * Get the WikiWho data for a given wiki page.
+	 *
+	 * @param {string} wikiUrl URL of the wiki page that we want to analyze.
 	 * @return {Promise}
 	 */
 	getData( wikiUrl ) {
```

Comment:
We need to do somethign here to allow for repeated calls to this function, since it's called from `app.js` which is triggered every time the `WhoWroteThat` trigger button is clicked.

My thought is that we could cache this promise, and return it with all the needed data. The system is supposed to allow for toggling, so it's conceivable that a user will click the "Who Wrote That?" button multiple times (initially it will load, then the user can click the [x] on the InfoBarWidget which resets back to the previous content, and then click the "Who Wrote That?" trigger again -- at this point, we should return the existing data from the API rather than re-request it)

Every time, it will call `getData()` -- if we cache the promise, it will mean that the first time the promise will run, and then resolve or reject -- and whatever in `app.js` will then run the `.then` or `.fail` on.

Subsequent clicks will run `getData()` but the promise will be cached -- which means it's already resolved (or rejected) which is a no-op, returning the original data, and the `then` and `fail` will act the same way.

Alternatively, we cache the state ("finished" / "ready" / etc) and use `getData()` just as a marker for finishing the operation (which is somewhat what you're doing now, since you're storing the results in class variables that can be called, so you just need to make sure they're ready.)

Either way, we have to make sure that `getData()` doesn't produce unnecessary API calls on subsequent clicks.

If we use a flag technique, then you can fake the return if the flag has already settled. For example,

```
getData( wikiUrl ) {
    if ( this.successful ) {
         return $.Deferred.resolve();
    } else if ( this.error ) {
         return $.Deferred.reject( this.error );
    }
    // The rest of the code, actually fetching
}
```

Otherwise, you should cache the promise and return it if it exists; something like
```
getData( wikiUrl ) {
   if ( this.promise ) {
         return this.promise;
   }
   this.promise = $.getJSON( ... ).then( .... );

   return this.promise;
}
```

Reply:
I've gone with promise-caching approach.